otcextensions/tests/functional/osclient/nat/v2/test_dnat_rule.py

```diff
@@ -0,0 +1,199 @@
+    ROUTER_ID = None
+    NETWORK_ID = None
+    FLOATING_IP_ID = None
+    PORT_ID = None
+
+    NAT_NAME = 'os-cli-test-' + UUID
+    NAT_ID = None
+    DNAT_RULE_ID = None
+
+    def test_01_nat_dnat_rule_list(self):
+        self.openstack(
+            'nat dnat rule list -f json '
+        )
+
+    def test_02_nat_dnat_rule_list_filters(self):
+        self.openstack(
+            'nat dnat rule list '
+            '--limit 1 --id 2 '
+            '--project-id 3 '
+            '--port-id 4 '
+            '--private-ip 5 '
+            '--internal-service-port 6 '
+            '--floating-ip-id 7 '
+            '--floating-ip-address 8 '
+            '--external-service-port 9 '
+            '--status 10 '
+            '--nat-gateway-id 11 '
+            '--protocol tcp '
+            '--admin-state-up true '
+        )
+
+    def test_03_nat_dnat_rule_create(self):
+        self._create_nat_gateway()
+        self.assertIsNotNone(self.PORT_ID)
+        self.assertIsNotNone(self.FLOATING_IP_ID)
+        json_output = json.loads(self.openstack(
+            'nat dnat rule create '
+            '--nat-gateway-id {nat_gateway_id} '
+            '--floating-ip-id {floating_ip_id} '
+            '--protocol {protocol} '
+            '--internal-service-port 80 '
+            '--external-service-port 80 '
+            '--private-ip {private_ip} '
+            '-f json'.format(
+                nat_gateway_id=self.NAT_ID,
+                protocol='TCP',
+                private_ip='192.168.0.3',
+                floating_ip_id=self.FLOATING_IP_ID)
+        ))
+        self.assertIsNotNone(json_output)
+        TestDnatRule.DNAT_RULE_ID = json_output['id']
+
+    def test_04_nat_dnat_rule_list_by_id(self):
+        self.assertIsNotNone(self.DNAT_RULE_ID)
+        json_output = json.loads(self.openstack(
+            'nat dnat rule list -f json '
+            '--id ' + self.DNAT_RULE_ID
+        ))
+        self.assertIsNotNone(json_output)
+        self.assertEqual(next(iter(json_output))['Id'], self.DNAT_RULE_ID)
+        self.assertEqual(
+            next(iter(json_output))['Nat Gateway Id'], self.NAT_ID)
+
+    def test_05_nat_dnat_rule_list_by_nat_id(self):
+        self.assertIsNotNone(self.DNAT_RULE_ID)
+        json_output = json.loads(self.openstack(
+            'nat dnat rule list -f json '
+            '--nat-gateway-id ' + self.NAT_ID
+        ))
+        self.assertIsNotNone(json_output)
+        self.assertEqual(
+            next(iter(json_output))['Nat Gateway Id'], self.NAT_ID)
+
+    def test_06_nat_dnat_rule_show(self):
+        self.assertIsNotNone(self.DNAT_RULE_ID)
+        json_output = json.loads(self.openstack(
+            'nat dnat rule show '
+            ' -f json ' + self.DNAT_RULE_ID
+        ))
+        self.assertIsNotNone(json_output)
+        self.assertEqual(json_output['id'], self.DNAT_RULE_ID)
+
+    def test_07_nat_dnat_rule_delete(self):
+        self.addCleanup(self._delete_nat_gateway)
```

Comment:
addCleanup should be invoked immediately after creating smth, it will be executed at the very end though - you can't trust the execution order (never do this)

Reply:
Resolved